### PR TITLE
[stable/fluent-bit] Revert to og indentation... or Refactor configuration

### PR DESCRIPTION
**Pull request by naseemkullah**
#### What this PR does / why we need it:

TL;DR: indentation is wrong, as these values are meant for extra ENTIRE sections rather than extra params per already exists sections.

Longer version:
@Towmeykaw originally added this for it's intended purpose of adding extra entire sections (e.g. a 2nd output)
I misinterpreted this as adding extra parameters for existing sections (e.g. adding `Replace_Dots On` within the ElasticSearch output section) and changed the indentation accordingly.

This PR will revert this back to the original intention.

As for adding extra parameters to already existing sections, well my particular case will be resolved via https://github.com/helm/charts/pull/8659/ and so I am waiting on that.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

**Review by max-lobur — APPROVED** [review]
👍 👍 thanks for reverting this. 
Prev patch broke our process as well

**Comment by max-lobur** [comment]
@naseemkullah you can add extra params to existing section using extra spaces in your var files, e.g.
```
extraInputs: |-
    Buffer_Max_Size 512k
```
yml will preserve 4 spaces before the option, and you get 4 from template, total 8, so it will get directly into config of the previous section

**Comment by naseemkullah** [comment]
Thanks for pointing that out @max-lobur !!

Regarding > 7a0265e#diff-9a7261a0279fb006f57a62fb2a9898fa, it allows us to replace an already existing option, rather than add one, e.g.:

If I were to add:
```
extraInputs: |2
      Mem_Buf_Limit 50KB
```

Would we not end up with 2 lines of Mem_Buf_Limit since the Tail input comes with one already?

**Comment by max-lobur** [comment]
@naseemkullah you're right, I didn't notice I have 2 same options in input :) I guess it takes the last, at least I see no errors

**Comment by max-lobur** [comment]
@naseemkullah I'd remove any fine-grained settings and move them to `extraInputs` etc. Config complexity always grows, it doesn't make sense to repeat that in a helm chart.

**Comment by naseemkullah** [comment]
@max-lobur Agreed!! I would only add settings that make this chart work out of the box for MOST cases.

I feel that `Replace_Dots On` is very important (since many of us will run  workloads with `app` labels and  others with `app.kubernetes.io/name` labels) and being able to configure `Mem_Buf_Limit` to be 50KB (or something much lower than 5MB), 5MB does not work for me at all. I would have set the default to be 50KB but I guess 5MB works for other people, having it configurable is a happy compromise I guess? Having it twice in the configmap feels weird, even if only second value is picked up. IMHO my 2 cents

**Comment by naseemkullah** [comment]
bump

**Comment by kfox1111** [comment]
See this thread: https://github.com/helm/charts/pull/9554
especially my last comment in it. I do not think reverting the indent is a good idea. We're getting an unwealdy amount of config options and have no way to test for breakages. Lets close this. If you need more config sections, you can always upload a configmap?

**Comment by Towmeykaw** [comment]
The reason I created it this way is that its easier to manage everything with helm than having some resources created outside when I have many clusters running. But I guess I will just keep maintaining my own chart.

**Comment by naseemkullah** [comment]
That's fine by me @kfox1111, but please consider the following:

SO I understand from the thread that "lets just cover the [input][filter][output] use case, and use your own configmap if you must differ."

In this case we keep the indentation at 4, but we should rename the values 
from: `extraInputs`  `extraFilters`   `extraOutputs`  
to: `extraInputParams` `extraFilterParam` and `extraOutputParams`

CCing @max-lobur and @Towmeykaw to whom I apologize for making this change in the first place

**Comment by naseemkullah** [comment]
I must admit that it is nice to manage the config entirely through the helm chart, so limiting to 1 input 1 filter and 1 output will alienate some people and potentially decrease contribution to upstream. MY two cents @kfox1111

**Comment by kfox1111** [comment]
The problem is, everyone wants to manage everything in values. and we end up with an unmaintainable mess of templates. This puts load on the reviewers which can't keep up with contributions. So one way or another someone will be disapointed I'm afraid.

Helm's subchart concept should allow you to just add your own configmap, and use a requires to nest this chart. then you don't have to maintain your own chart. just your own configmap. I think this is a good compromise? You get the level of customization you want, and we all still share as much as we can.

As for what to call the options, I'm a bit nervous also continuing to rename options over and over. we're going to keep breaking users if we do. Do we have a handle on how many users are using the existing behavior?

**Comment by Towmeykaw** [comment]
Yeah that seems like a good approach to me.

**Comment by max-lobur** [comment]
This revert would give me ability to manage `extraInputs` and options as  I mentioned here: https://github.com/helm/charts/pull/9008#issuecomment-439433178

He reverts the change on purpose, he unintentionally broke the original idea and now wants to fix it.
 
@kfox1111 If you wanna stop accepting config changes here I agree, but let's do this after everything is returned to normal, original look.

**Comment by kfox1111** [comment]
One of the issues I had with the previous approach is this:
extraInputs: |2
      Mem_Buf_Limit 50KB

We're taking toml encoded strings, craming them in yaml, then adding yaml tweaks to add white space to the lines, and assuming the whitespace in the configmap matches with what the user specifies.

This feels very fragile. It would prevent us from making whitespace changes in the config file. Its hard enough to maintain the options that are coming in now without breaking things, let alone having options that depend on what whitespace the user has specified.

This feels very wrong to me to be encouraging.

Then, today I found out that filters can be before and after other non filters. so the previous behavior isn't generic enough. Just enough to cover your current desire. It would be likely then someone else could come along in a month and ask for another config option to handle the more general yet case.

Can we find a different way of supporting your use case that doesn't have these problems?

**Comment by max-lobur** [comment]
@kfox1111 have separate `extraInputOptions` (indent 8) and `extraInputs` (indent 4) in the same time :)

**Comment by max-lobur** [comment]
I'm doing my own template right now to overcome this anyways.
Either naming needs to be fixed, or indent. It was very confusing to upgrade and discover that the helm var has changed it's purpose

**Comment by max-lobur** [comment]
Also docs need to be updated if you wanna keep 8. https://github.com/helm/charts/blob/master/stable/fluent-bit/README.md
Still states its for extra config sections, not options

**Comment by naseemkullah** [comment]
To summarize, options include:

1. Leave indentation and value names as they are, but at the very least update the README

2. Leave indentation as is, but adjust value names to `extraInputParams` `extraFilterParams` `extraOutputParams` , and update README

3. Have both `extraInputs` `extraFilters` `extraOuputs` and  `extraInputParams` `extraFilterParams` `extraOutputParams` with 8 and 4 indentation relatively, and update the README

Right?

**Comment by kfox1111** [comment]
Yeah, trying to figure out a way without adding lots of new options if possible....

I wasn't involved in the original pr that added them in the first place. I would likely have pushed back if I saw it because of these sorts of issues. So, I am sorry it broke on you. But I'm trying to fix it as best as I can without breaking others too or making an unmaintainable mess. It is possible others are relying on the current behavior.

In light of the ordering behavior mentioned before with sections, the old names don't make a lot of sense to me either.

extraInputs is really, drop this chunk of text in between the main INPUT and FILTER. You could just as easily put a new filter section in extraInputs if you wanted it to show up before the inline filter.

So, extraInputs being indent 4 is not a very good name for it. Not sure what it should be called. Any ideas?

**Comment by naseemkullah** [comment]
extraSections?

**Comment by naseemkullah** [comment]
>extraInputs is really, drop this chunk of text in between the main INPUT and FILTER. You could just as easily put a new filter section in extraInputs if you wanted it to show up before the inline filter

Very true, if we do not decide on an intention (entire Section or added param) and just leave it as a tweak-able chunk of text, I do not see a good name for it. So we should probably set our intention and it would be extra sections, since that's what the populous seem to want, in which case indentation should probably be reverted to 8.

Now that you've mentioned that placement is not important, we could potentially convert `extraInputs` extraFilters` and `extraOutputs` into `extraSections` though.

As for adding parameters to the out-of-the-box Input, Filter and Output. At this point provided #8659 goes through, I no longer have a need FWIW, but that flexibility would be nice to have, and easily addable as `extraInputParams` `extraFilterParams` and `extraOutputParams` with the indentation of 4.

For good measure we can go up a major version (to 17.0.0) to demonstrate that there are breaking changes. My 2 cents

**Comment by naseemkullah** [comment]
Actually never mind I see that according to your slack convo with @donbowman order does matter, I will remove my previous comment about `extraSections`.

**Comment by kfox1111** [comment]
We've got all these cases I think:
1. add lines to the end of the existing INPUT section
2. add lines to the end of the existing FILTER section
3. add lines to the end of the existing OUTPUT section
4. add raw lines before INPUT
5. add raw lines between INPUT and FILTER
6. add raw lines between FILTER and OUTPUT
7. add raw lines after OUTPUT

If we are to support them all, they need to be called something that is easily understandable by a user, so they can set them. If we can't come up with a good names, then its unlikely users will use the features. I'm really at a loss for what to call 4-7. raw[0-3] or raw[1-4] seems wrong. beforeInput, afterOutput maybe.... betweenInputAndFilter? betweenFilterAndOutput?

```yaml
extraConfig
  lines:
    input: |
      line to go into input
  raw:
    betweenInputAndFilter: |
      [FILTER]
      onefilter
    betweenFilterAndOutput: |
      [FILTER]
      otherfilter
```
?

**Comment by max-lobur** [comment]
@kfox1111 I like it.
```
extraConfig
  options:
    input: |
      line to go into input
    filter: ...
  sections:
    afterInput: |
      [INPUT]
      onefilter
    afterFilter: |
      [FILTER]
      otherfilter
    afterOutput: |
      [OUTPUT]
       ....
```

**Comment by max-lobur** [comment]
Any user will look to template anyways, before relying on options, so naming `afterInput` and so on is not super crit. Can call it just `input`, `filter` and so on, main thing is that they should not change its purpose over time

**Comment by max-lobur** [comment]
Grafana handles ini in interesting way:
https://github.com/helm/charts/blob/master/stable/grafana/templates/configmap.yaml#L14-L20

Basically doing bare conversion from yaml to ini. Might work for all of us.

**Comment by kfox1111** [comment]
Input is unclear as it doesn't say if its before or after the existing input. same with filter/output. I have been forced many times to look at the template to make sense of values, so I kind of get what you are saying. But, I do consider it a broken chart when you must. Ideally a chart should be able to be launched via a gui, where all it knows about is the values.

ini mapping works, as it is a pure subset of yaml.

sax parsed toml on the other hand, is not. So coming up with a sane mapping is tricky/very time consuming/error prone. I for sure don't have time to come up with a good mapping. If you do, that would be great.

Better yet might be to update fluent-bit to import yaml as a config option. The mapping would then be trivial for the chart.

For now, uploading your own configmap is just as flexible without the large development effort. Yes, it does put some extra effort on an end user if they want to slip in extra filters. But I think your already under the 20% use case for this.

Fluent-bit really should not be filtering much IMHO. Its lightweight log shipper. The more processing it does, the more resource its taking away from the node doing work. It should just ship logs to fluentd for doing complex log processing before storage. Keep the shipping and processing layers separate?

**Comment by kfox1111** [comment]
Actually, I'm not sure that it is toml rather then a custom config file format. I may have been mistaken about that part.

**Comment by naseemkullah** [comment]
Are we all ok with using the grafana (ini mapping) approach? 
Can that work even if it is toml/some other config file format?

**Comment by kfox1111** [comment]
the problem is coming up with an appropriate mapping.

ini's easy, as it is a 2 level only dict of dicts of strings.
```yaml
sectionname:
  keyname: value
```

What is the mapping from fluent-bit's config file format?
the sections are not unique, so using a dict wouldn't work. so closest I can think of is an array of dicts of strings?

raw:
- type: FILTER
  data: |
    foo
- type: FILTER
  data: |
    bar

helm can't merge array values together though via values, so it couldn't merge with the existing stuff without going back to having the inbetween values sections... So kind of defeats the purpose.

**Comment by naseemkullah** [comment]
Ah I see, I see... 
So, what do we do in the end?

**Comment by kfox1111** [comment]
ok... I have one more idea. Its kind of crazy, but might work.
So, we leave the three extra*'s in place as they let you slide in lines to the existing sections.

We modify the configmap to split the current config file into 5 files instead of the 1 config file.
4 of them will be fluent-bit-{{$sectionName}}.conf and contain just the contents of each of the [{{$sectionName}}] that exists in config.yaml already.

the 5th one would be the new fluent-bit.conf and it will contain the raw contents of:
rawConfig

Then we set in values.yaml:
```yaml
rawConfig: |
 @INCLUDE fluent-bit-service.conf
 @INCLUDE fluent-bit-input.conf
 @INCLUDE fluent-bit-filter.conf
 @INCLUDE fluent-bit-output.conf
``` 

Then, anyone can customize anything by setting their own rawConfig and putting whatever they want, wherever they want. Inbetween input and output, before service, and evel leave out includes like input if they want their whole own one.

We're really starting to twist ourselves in knots though.

**Comment by naseemkullah** [comment]
Clever! A little bloated but very flexible. :)
It's up to you guys to hash it out in the end, I'm just awaiting orders at this point.

**Comment by Towmeykaw** [comment]
I think that would be a nice solution.

**Comment by kfox1111** [comment]
@naseemkullah sounds like there is consensus. Do you have time to take a stab at implementing it?

**Comment by naseemkullah** [comment]
Great to hear! yes @kfox1111 with pleasure

**Comment by naseemkullah** [comment]
@kfox1111 @Towmeykaw @max-lobur 
Here's a first stab at it, please let me know what you think.

Please note I did not implement any `sectionNames` list logic, if you would like me to please elaborate as to how.

Please also confirm, if someone would like to add an option, suppose `Replace_Dot On` in the output section, they would put this in their values file?
```
[...]
rawConfig: |-
  @INCLUDE fluent-bit-service.conf
  @INCLUDE fluent-bit-input.conf
  @INCLUDE fluent-bit-filter.conf
  @INCLUDE fluent-bit-output.conf
      Replace_Dots On
[...]
```

Is there a way to see the computed resulting config clearly, particularly the added option appended to the output section?

**Comment by naseemkullah** [comment]
After getting @donbowman 's consent whose PR #8659 has been taking >1 month to be merged I'm taking this opportunity to sneak in the `Replace_Dots` option as well. TIA

**Comment by naseemkullah** [comment]
I've updated the chart to `1.0.0` and added an upgrading section in the README.

**Review by max-lobur — APPROVED** [review]
works for me

**Comment by donbowman** [comment]
we still have the issue that subPath is used. subPath configmaps don't update on changes, so caveat to the user they have to remove and replace rather than 'helm upgrade'.

**Comment by kfox1111** [comment]
looking pretty good. can you add back extra*? Right now its the only way to append extra rules into the existing INPUT/FILTER/OUTPUT section without guessing the amount of white space. What should we call it though?
maybe:
```yaml
extraConfig:
  input: |
     some input lines
  filter: |
     some filter lines
  output: |
     some output lines
```
?

We do need to get  @donbowman's patch in too before 1.0. should we merge that into this pr, or release this as a 0.17 and then do 1.0 later?

**Comment by kfox1111** [comment]
Oh, nice. looks like @edsiper got through an owners file. so reviews should be easier now.

**Comment by naseemkullah** [comment]
Thanks everyone for the feedback and sorry for all the forced pushes and failed CIs. Please review the following changes:

- Merged @donbowman's patch into this commit.
- Added back `extra*` as extraEntries (this is the vocabulary used on fluent-bit website)
- Input path configurable to address issue [#9760](https://github.com/helm/charts/issues/9670)

File: stable/fluent-bit/templates/config.yaml [hunk 1]
```diff
@@ -27,7 +27,7 @@ data:
   fluent-bit-input.conf: |-
     [INPUT]
         Name             tail
-        Path             /var/log/containers/*.log
+        Path             {{ .Values.input.tail.path }}
```
Comment (nvoskuilen) [inline stable/fluent-bit/templates/config.yaml:30]: Thanks!

Comment (naseemkullah) [reply stable/fluent-bit/templates/config.yaml:30]: np! :)

**Comment by kfox1111** [comment]
Thanks all for working through this. I think the end result turned out very well. :)

**Comment by naseemkullah** [comment]
Agreed! Thanks everybody!
